File: src/go/problems/Subsets_test.go
// https://leetcode.com/problems/subsets/

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type Subsets struct{}

func TestSubsets(t *testing.T) {
	gen := core.TestSuite[Subsets]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]int{1,2,3}).Result([][]int{{},{1},{2},{1,2},{3},{1,3},{2,3},{1,2,3}})
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{0}).Result([][]int{{},{0}})
	}).Run(t)
}

func (Subsets) Solution(nums []int) [][]int {
	result := [][]int{}

	search(nums, 0, []int{}, &result)

	return result
}

func search(nums []int, index int, temp []int, result *[][]int) {
	tempCopy := make([]int, len(temp))
	copy(tempCopy, temp)
	*result = append(*result, tempCopy)

	for i := index; i < len(nums); i++ {
		temp = append(temp, nums[i])
		search(nums, i+1, temp, result)
		temp = temp[:len(temp)-1]
	}
}

File: src/python/problems/subsets_test.py
#https://leetcode.com/problems/subsets/

from core.problem_base import *

class Subsets(ProblemBase):
    def Solution(self, nums: list[int]) -> list[list[int]]:
        result = []

        def search(index, temp):
            result.append(temp.copy())

            for i in range(index, len(nums)):
                temp.append(nums[i])
                search(i + 1, temp)
                temp.pop()

        search(0, [])

        return result

if __name__ == '__main__':
    TestGen(Subsets) \
        .Add(lambda tc: tc.Param([1,2,3]).Result([[],[1],[1,2],[1,2,3],[1,3],[2],[2,3],[3]])) \
        .Add(lambda tc: tc.Param([0]).Result([[],[0]])) \
        .Run()
